### PR TITLE
version: make `formula_optionally_versioned_regex` public

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

There are over 20 formulae that make use of a similar regex that could
be replaced with a call to `formula_optionally_versioned_regex`, so I'd
like to be able to use this there.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]

File: Library/Homebrew/version.rb [hunk 1]
```diff
@@ -13,6 +13,7 @@ class Version
 
   include Comparable
 
+  # @api public
```
Comment (MikeMcQuaid) [inline Library/Homebrew/version.rb:16]: If it's public: might want to document it too?

Comment (carlocab) [reply Library/Homebrew/version.rb:19]: Thanks; added documentation.

**Review by MikeMcQuaid — COMMENTED** [review]
Code looks good!

> There are over 20 formulae that make use of a similar regex that could
> be replaced with a call to `formula_optionally_versioned_regex`, so I'd like to be able to use this there.

Can you link to one of them? Would love to figure out if there's a better API for them than a method that returns a regex 🤔

**Comment by carlocab** [comment]
Here are a few:

https://github.com/Homebrew/homebrew-core/blob/acd5bd472f377518006d53363ffab452c8a2047b/Formula/boost-python3.rb#L27-L30

https://github.com/Homebrew/homebrew-core/blob/acd5bd472f377518006d53363ffab452c8a2047b/Formula/byobu.rb#L34-L37

https://github.com/Homebrew/homebrew-core/blob/acd5bd472f377518006d53363ffab452c8a2047b/Formula/pillow.rb#L35-L39

https://github.com/Homebrew/homebrew-core/blob/acd5bd472f377518006d53363ffab452c8a2047b/Formula/vapoursynth-ocr.rb#L47-L50

https://github.com/Homebrew/homebrew-core/blob/acd5bd472f377518006d53363ffab452c8a2047b/Formula/pybind11.rb#L27-L30

**Comment by MikeMcQuaid** [comment]
@carlocab feels like a higher-level DSL might be a better fit; something that allows you to get all the deps that match e.g. `python` or `Formula["python"].versioned_formulae` by passing in `"python"` to a `Formula` method. Thoughts?

**Comment by carlocab** [comment]
Yes, I agree. Let me have a think about the right way to do it.

**Pull request closed without merging**